src/snowflake/cli/api/utils/graph.py

```diff
@@ -0,0 +1,79 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+
+class VisitStatus(Enum):
+    VISITING = 1
+    VISITED = 2
+
+
+@dataclass
+class Node(Generic[T]):
+    key: str
+    data: T
+    neighbors: set[Node[T]] = field(default_factory=set)
+
+    def __hash__(self):
+        return hash(self.key)
+
+    def __eq__(self, other):
+        return self.key == other.key
+
+
+class Graph(Generic[T]):
+    def __init__(self):
+        self._graph_nodes_map: dict[str, Node[T]] = {}
+
+    def get(self, key: str) -> Node[T]:
+        if key in self._graph_nodes_map:
+            return self._graph_nodes_map[key]
+        raise KeyError(f"Node with key {key} not found")
+
+    def get_all_nodes(self) -> set[Node[T]]:
+        return set(self._graph_nodes_map.values())
+
+    def add(self, node: Node[T]) -> None:
+        if node.key in self._graph_nodes_map:
+            raise KeyError(f"Node key {node.key} already exists")
+        self._graph_nodes_map[node.key] = node
+
+    def add_directed_edge(self, from_node_key: str, to_node_key: str) -> None:
+        from_node = self.get(from_node_key)
+        to_node = self.get(to_node_key)
+        from_node.neighbors.add(to_node)
+
+    @staticmethod
+    def _dfs_visit(
+        nodes_status: dict[str, VisitStatus],
+        node: Node[T],
+        visit_action,
+        on_cycle_action,
```

Comment:
these are both (required) callable params, let's add type annotations. Same for the dfs entry point.